# Day 1
The newly-improved calibration document consists of lines of text; each line originally contained a specific calibration value that the Elves now need to recover. On each line, the calibration value can be found by combining the first digit and the last digit (in that order) to form a single two-digit number.

For example:
```
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
```
In this example, the calibration values of these four lines are 12, 38, 15, and 77. Adding these together produces 142.

Consider your entire calibration document. What is the sum of all of the calibration values?

In [10]:
import re

In [ ]:
file_path = '../data/day1.txt'
with open(file_path, 'r') as file:
    lines = [line.strip() for line in file]

len(lines)

## Part 1

In [26]:
def remove_non_digits(s: str):
    return re.sub(r'\D', '', s)

def extract_digits(s: str):
    # no empty lines according to the description
    if len(s) == 1:
        s = s*2
    else:
        s = s[0] + s[-1]
    return int(s)

In [31]:
res = sum([extract_digits(remove_non_digits(s)) for s in lines])
res

54159

## Part 2

In [77]:
from typing import List

In [109]:
def find_overlapping_matches(pattern, text):
    # Use a lookahead assertion to find overlapping matches
    return [match.group(1) for match in re.finditer(f'(?=({pattern}))', text)]

def extract_allowed_values(original_string, allowed_values):
    # concat literal strings
    pattern = '|'.join(re.escape(value) for value in allowed_values)
    pattern += r'|\d'
    # return re.findall(pattern, original_string)
    return find_overlapping_matches(pattern, original_string)

In [112]:
def convert_spelled(l: List[str], mapping):
    spelled = list(map(lambda x: mapping[x] if x in mapping else x, l))
    return spelled

In [113]:
def extract_first_last(s: List[str]):
    # no empty lines according to the description
    if len(s) == 1:
        s = s[0]*2
    else:
        s = s[0] + s[-1]
    return int(s)

In [114]:
def extract_calibration_value(s, mapping):
    allowed_values = extract_allowed_values(s, mapping.keys())
    digits = convert_spelled(allowed_values, mapping)
    calibration_value = extract_first_last(digits)
    return calibration_value

In [115]:
for s in lines:
    print(s)
    print(extract_calibration_value(s, spelled_digits_mapping))

eightqrssm9httwogqshfxninepnfrppfzhsc
89
one111jxlmc7tvklrmhdpsix
16
bptwone4sixzzppg
26
ninezfzseveneight5kjrjvtfjqt5nineone
91
58kk
58
5b32
52
1dtwo
12
six7two7sixtwo78
68
mvhsixpptztjh13sixthree2
62
six1bqqvrxndt
61
fourmk5grmqone944nbvtj
44
twofiveqxfivezpkvfvxt5eightjhnpl
28
fpfqp7three7
77
scmlf76ninegjzjkj97two
72
fivetkhfnnx22
52
sevenxvbcbsvxr7eighttwo
72
1hvhqqmrs1bgttshthg6
16
4bvnccbdh4onefztdrpq62vvbnvpxxvgrngnfjgfk
42
653spgrvd
63
sixctlhkjmmxh2fourfivenine37
67
229mjp3txmqsxxqdbnnnbrtrcctgzseven
27
jfourdbpcjc39bhglgnine
49
bvnltxdmsp7twoxzpdjdvkxeight4twothree
73
jlvcdrkhzh8seven3
83
418oneeight
48
53flcrlvqdeight84frmdcsixchcbc
56
114sixone1eight2
12
xrbtzbklqsl11
11
bhfhszrhzgrhsfd2threeseventwosevenoneseven
27
four9one
41
5p
55
twovhjpdxmcxshnhv5vs
25
qkkqeightcxcltnn7one9pmhlmvsxnine
89
4cbptmvp1
41
84xgm
84
bzsmqhkrdtdmhhjgrjsdfour1ninetwo61
41
onetwoeightgflhlgksevennine7two6
16
mbjhkhfour6
46
8cvqk6eightonethree1
81
qhbllbnlkr3rcsmjvztgd
33
18eight4
14
hhc6onegvk

In [116]:
spelled_digits_mapping = {'one': '1', 'two': '2', 'three': '3', 'four': '4', 'five': '5', 'six': '6', 'seven': '7', 'eight': '8', 'nine': '9'}

res = sum([extract_calibration_value(s, spelled_digits_mapping) for s in lines])
res

53866

In [97]:
test = ['two1nine', 'eightwothree', 'abcone2threexyz', 'xtwone3four', '4nineeightseven2', 'zoneight234', '7pqrstsixteen']
[extract_calibration_value(s, spelled_digits_mapping) for s in test]
sum([extract_calibration_value(s, spelled_digits_mapping) for s in test])

281

# Day 2

## Part 1
You play several games and record the information from each game (your puzzle input). Each game is listed with its ID number (like the 11 in Game 11: ...) followed by a semicolon-separated list of subsets of cubes that were revealed from the bag (like 3 red, 5 green, 4 blue).

For example, the record of a few games might look like this:
```
Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
```
In game 1, three sets of cubes are revealed from the bag (and then put back again). The first set is 3 blue cubes and 4 red cubes; the second set is 1 red cube, 2 green cubes, and 6 blue cubes; the third set is only 2 green cubes.

The Elf would first like to know which games would have been possible if the bag contained only 12 red cubes, 13 green cubes, and 14 blue cubes?

In the example above, games 1, 2, and 5 would have been possible if the bag had been loaded with that configuration. However, game 3 would have been impossible because at one point the Elf showed you 20 red cubes at once; similarly, game 4 would also have been impossible because the Elf showed you 15 blue cubes at once. If you add up the IDs of the games that would have been possible, you get 8.

Determine which games would have been possible if the bag had been loaded with only 12 red cubes, 13 green cubes, and 14 blue cubes. What is the sum of the IDs of those games?

In [1]:
load = {'red': 12, 'green': 13, 'blue': 14}

file_path = '../data/day2.txt'
with open(file_path, 'r') as file:
    lines = [line.strip() for line in file]

len(lines)

100

In [9]:
l = lines[50]
l

'Game 51: 4 red, 18 green, 1 blue; 15 green; 3 blue, 16 red, 17 green; 9 blue, 10 red, 13 green; 11 green, 14 red; 1 green, 7 blue, 1 red'

In [41]:
def parse_line(l):
    s_without_game = l.split("Game ")[1]
    game_id, sets_str = s_without_game.split(': ')
    return int(game_id), sets_str.split('; ')

parse_line(lines[50])

(51,
 ['4 red, 18 green, 1 blue',
  '15 green',
  '3 blue, 16 red, 17 green',
  '9 blue, 10 red, 13 green',
  '11 green, 14 red',
  '1 green, 7 blue, 1 red'])

In [45]:
import re
pattern = r"(\d+)\s(" + "|".join(load.keys()) + ")"

def set2dict(input):
    matches = re.findall(pattern, input)
    return {k: int(v) for v, k in matches}

set2dict(sets[1])

{'red': 9, 'green': 6, 'blue': 4}

In [46]:
def check_correct(input, bag):
    for k, v in input.items():
        if bag[k] < v:
            return False
    return True

def check_game(sets, load):
    for s in sets:
        set_dict = set2dict(s)
        if not check_correct(set_dict, load):
            return False
    return True

In [47]:
sum = 0
for l in lines:
    game_id, sets = parse_line(l)
    if check_game(sets, load):
        sum += game_id

In [48]:
sum

2600

## Part 2

The Elf says they've stopped producing snow because they aren't getting any water! He isn't sure why the water stopped; however, he can show you how to get to the water source to check it out for yourself. It's just up ahead!

As you continue your walk, the Elf poses a second question: in each game you played, what is the fewest number of cubes of each color that could have been in the bag to make the game possible?

Again consider the example games from earlier:
```
Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
```

* In game 1, the game could have been played with as few as 4 red, 2 green, and 6 blue cubes. If any color had even one fewer cube, the game would have been impossible.
* Game 2 could have been played with a minimum of 1 red, 3 green, and 4 blue cubes.
* Game 3 must have been played with at least 20 red, 13 green, and 6 blue cubes.
* Game 4 required at least 14 red, 3 green, and 15 blue cubes.
* Game 5 needed no fewer than 6 red, 3 green, and 2 blue cubes in the bag.

The power of a set of cubes is equal to the numbers of red, green, and blue cubes multiplied together. The power of the minimum set of cubes in game 1 is 48. In games 2-5 it was 12, 1560, 630, and 36, respectively. Adding up these five powers produces the sum 2286.

For each game, find the minimum set of cubes that must have been present. What is the sum of the power of these sets?

In [49]:
from collections import defaultdict
from itertools import chain

In [54]:
def calculate_set_power(sets):
    set_dicts = [set2dict(s) for s in sets]
    combined_dict = defaultdict(list)

    # Chaining all key-value pairs from all dictionaries and accumulating values
    for key, value in chain(*[d.items() for d in set_dicts]):
        combined_dict[key].append(value)

    # Converting back to a regular dictionary for final output (optional)
    combined_dict = dict(combined_dict)
    power = 1
    for v in combined_dict.values():
        power *= max(v)
    return power

In [56]:
sum = 0
for l in lines:
    game_id, sets = parse_line(l)
    power = calculate_set_power(sets)
    sum += power
sum

86036

# Day 3

## Part 1

The engineer explains that an engine part seems to be missing from the engine, but nobody can figure out which one. If you can add up all the part numbers in the engine schematic, it should be easy to work out which part is missing.

The engine schematic (your puzzle input) consists of a visual representation of the engine. There are lots of numbers and symbols you don't really understand, but apparently any number adjacent to a symbol, even diagonally, is a "part number" and should be included in your sum. (Periods (.) do not count as a symbol.)

Here is an example engine schematic:
```
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
```
In this schematic, two numbers are not part numbers because they are not adjacent to a symbol: 114 (top right) and 58 (middle right). Every other number is adjacent to a symbol and so is a part number; their sum is 4361.

Of course, the actual engine schematic is much larger. What is the sum of all of the part numbers in the engine schematic?

In [2]:
file_path = '../data/day3.txt'
with open(file_path, 'r') as file:
    lines = [line.strip() for line in file]

len(lines), len(lines[0])

(140, 140)

everything apart from digits and dots are symbols

conditions:
1. first digit after a symbol
2. last digit before a symbol
3. above and below lines have any symbol between (inclusively) first digit index - 1 and last digit index + 1

the first two rules can be checked in one run over the line (if I do not solve this with regex). the second is a function call for check if any

In [3]:
import re

def contains_only_digits_and_dots(s):
    return re.search("[^\d.]", s) is None

contains_only_digits_and_dots('3#')

False

In [4]:
contains_only_digits_and_dots('3')

True

In [5]:
def yield_integer_sequences(s):
    # both indices are on the exact positions
    pattern = r'\d+'
    for match in re.finditer(pattern, s):
        yield (match.start(), match.end()-1, match.group())

string = '.855.442.'
for start, end, value in yield_integer_sequences(string):
    print(f"Integer sequence from index {start} to {end}: {string[start:end+1]} (string) or {value} (value)")

Integer sequence from index 1 to 3: 855 (string) or 855 (value)
Integer sequence from index 5 to 7: 442 (string) or 442 (value)


In [6]:
contains_only_digits_and_dots(lines[0][0])

True

In [7]:
total_lines = len(lines)
total_lines

140

In [8]:
def other_line_contains_any_symbol(l, start, end):
    # start, end - original line indixes
    line_len = len(l)
    if start > 0:
        start -= 1
    if end < (line_len - 1):
        end += 1
    # print(f"checking sequence '{l[start:end+1]}'")
    res = not contains_only_digits_and_dots(l[start:end+1])
    # if res:
    #     print(f"this context works: '{l[start:end+1]}'")
    return res

# original                      ..11
other_line_contains_any_symbol('...*', 2, 3)

True

In [9]:
def check_line_with_context(l, prev_line, next_line):
    part_numbers = []
    for start, end, value in yield_integer_sequences(l):
        line_len = len(l)
        if start != 0 and not contains_only_digits_and_dots(l[start - 1]):
            part_numbers.append(int(value))
            continue
        if end != (line_len - 1) and not contains_only_digits_and_dots(l[end + 1]):
            part_numbers.append(int(value))
            continue
        if prev_line and other_line_contains_any_symbol(prev_line, start, end):
            part_numbers.append(int(value))
            continue
        if next_line and other_line_contains_any_symbol(next_line, start, end):
            part_numbers.append(int(value))
            continue

    return part_numbers
            
check_line_with_context(lines[1], None, None)

[577, 465, 169]

In [10]:
check_line_with_context(l=lines[1], prev_line=lines[0], next_line=lines[2])

[21, 971, 577, 316, 465, 169]

In [13]:
total = 0
for line_number, l in enumerate(lines):
    if line_number > 0:
        prev_line = lines[line_number - 1]
    else:
        prev_line = None
    if line_number < total_lines - 1:
        next_line = lines[line_number + 1]
    else:
        next_line = None
    
    matched_numbers = check_line_with_context(l=l, prev_line=prev_line, next_line=next_line)
    print('line_number', line_number, 'matched_numbers:', matched_numbers, 'sum:', sum(matched_numbers))
    total += sum(matched_numbers)
        
total  

line_number 0 matched_numbers: [190, 520] sum: 710
line_number 1 matched_numbers: [21, 971, 577, 316, 465, 169] sum: 2519
line_number 2 matched_numbers: [881, 635, 577, 864, 873, 742, 714, 244] sum: 5530
line_number 3 matched_numbers: [602, 423, 939, 906, 899, 965, 309] sum: 5043
line_number 4 matched_numbers: [737, 294, 321, 337, 305, 153, 130, 638] sum: 2915
line_number 5 matched_numbers: [821, 544, 282, 165, 836, 112] sum: 2760
line_number 6 matched_numbers: [456, 281, 847, 448, 728, 833] sum: 3593
line_number 7 matched_numbers: [830, 34, 911, 702, 21, 279, 226] sum: 3003
line_number 8 matched_numbers: [192, 247, 40, 504] sum: 983
line_number 9 matched_numbers: [304, 938, 258, 679, 702, 742, 656, 645, 916, 263] sum: 6103
line_number 10 matched_numbers: [841, 168, 633, 758, 993, 410, 920, 195, 492] sum: 5410
line_number 11 matched_numbers: [988, 689, 554, 245, 794] sum: 3270
line_number 12 matched_numbers: [489, 866, 354, 567, 949, 969, 913, 10] sum: 5117
line_number 13 matched_numbe

531932

## Part 2

The missing part wasn't the only issue - one of the gears in the engine is wrong. A gear is any * symbol that is adjacent to exactly two part numbers. Its gear ratio is the result of multiplying those two numbers together.

This time, you need to find the gear ratio of every gear and add them all up so that the engineer can figure out which gear needs to be replaced.

Consider the same engine schematic again:
```
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
```
In this schematic, there are two gears. The first is in the top left; it has part numbers 467 and 35, so its gear ratio is 16345. The second gear is in the lower right; its gear ratio is 451490. (The * adjacent to 617 is not a gear because it is only adjacent to one part number.) Adding up all of the gear ratios produces 467835.

What is the sum of all of the gear ratios in your engine schematic?

There may be a nice solution if we had a dict of gear locations where elements are all adjacent part numbers. The solution based on the 1st part (based on numbers) will contain crazy conditions. 

So we base the solution on the gear and search for the adjacent part numbers.

1. The current line and neighbouring are considered
2. Asterisk is found. Store index to dict (line, line_index)
3. In-line adjacency check. Add to dict values list
4. Other line (both: top and bot) may have 1\2 diagonal neighbours or only 1 intersecting straght the same index
    * if index above\below the asterisk is a digit, find it's borders (both sides) and store to the dict values list. break
    * elif 
        * index above\below -1 is a digit, find it's border to the left and store to the dict values list. DON'T BREAK
        * index above\below +1 is a digit, find it's border to the right and store to the dict value list.

In [61]:
def find_number(l, idx, direction) -> str:
    if direction == 0:
        current_symbol = l[idx]
        if current_symbol.isdigit():  # there is a number crossing this idx
            return find_number(l=l, idx=idx, direction=-1) + current_symbol + find_number(l=l, idx=idx, direction=1)
    
    elif direction == -1:
        if idx <= 0:
            return ''
        
        left_symbol = l[idx-1]
        if left_symbol.isdigit():
            return find_number(l, idx-1, direction=direction) + left_symbol
        else:
            return ''
    
    elif direction == 1:
        if idx >= len(l) - 1:
            return ''
        
        right_symbol = l[idx+1]
        if right_symbol.isdigit():
            return right_symbol + find_number(l, idx+1, direction=direction)
        else:
            return ''

find_number('.....621.3', 3, 0)

In [62]:
def find_in_line(l, idx):
    left_in_line = find_number(l=l, idx=idx, direction=-1)
    right_in_line = find_number(l=l, idx=idx, direction=1)
    res = []
    if left_in_line: res.append(left_in_line)
    if right_in_line: res.append(right_in_line)
    return res

def find_in_other_line(l, idx):
    non_diag_number = find_number(l=l, idx=idx, direction=0)
    if non_diag_number:
        return [non_diag_number]
    else:
        return find_in_line(l=l, idx=idx)

In [63]:
def find_gears(l, prev_line, next_line, line_number, gears_dict):
    for idx, el in enumerate(l):
        if el == '*':
            gears_dict[(line_number, idx)] = []
            gears_dict[(line_number, idx)] += find_in_line(l, idx)
            if prev_line:
                gears_dict[(line_number, idx)] += find_in_other_line(l=prev_line, idx=idx)
            if next_line:
                gears_dict[(line_number, idx)] += find_in_other_line(l=next_line, idx=idx)
            
    return gears_dict

find_gears(lines[1], lines[0], lines[2], 1, {})

{(1, 72): ['520', '864'], (1, 119): ['465', '169']}

In [65]:
gears_dict = {}
for line_number, l in enumerate(lines):
    if line_number > 0:
        prev_line = lines[line_number - 1]
    else:
        prev_line = None
    if line_number < total_lines - 1:
        next_line = lines[line_number + 1]
    else:
        next_line = None
    
    gears_dict = find_gears(l=l, prev_line=prev_line, next_line=next_line, line_number=line_number, gears_dict=gears_dict)

In [70]:
lengths = [len(v) for v in gears_dict.values()]
min(lengths), max(lengths)

(1, 2)

thanks god, there are no cases with 3-member gears

In [73]:
total = 0
for k, v in gears_dict.items():
    if len(v) == 2:
        # print(k, v)
        total += int(v[0]) * int(v[1])

total

73646890

# Day 4

## Part 1
The Elf leads you over to the pile of colorful cards. There, you discover dozens of scratchcards, all with their opaque covering already scratched off. Picking one up, it looks like each card has two lists of numbers separated by a vertical bar (|): a list of winning numbers and then a list of numbers you have. You organize the information into a table (your puzzle input).

As far as the Elf has been able to figure out, you have to figure out which of the numbers you have appear in the list of winning numbers. The first match makes the card worth one point and each match after the first doubles the point value of that card.

For example:
```
Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53
Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19
Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1
Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83
Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36
Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11
```
In the above example, card 1 has five winning numbers (41, 48, 83, 86, and 17) and eight numbers you have (83, 86, 6, 31, 17, 9, 48, and 53). Of the numbers you have, four of them (48, 83, 17, and 86) are winning numbers! That means card 1 is worth 8 points (1 for the first match, then doubled three times for each of the three matches after the first).

Card 2 has two winning numbers (32 and 61), so it is worth 2 points.
Card 3 has two winning numbers (1 and 21), so it is worth 2 points.
Card 4 has one winning number (84), so it is worth 1 point.
Card 5 has no winning numbers, so it is worth no points.
Card 6 has no winning numbers, so it is worth no points.
So, in this example, the Elf's pile of scratchcards is worth 13 points.

Take a seat in the large pile of colorful cards. How many points are they worth in total?

In [52]:
file_path = '../data/day4.txt'
with open(file_path, 'r') as file:
    lines = [line.strip() for line in file]

len(lines), len(lines[0])

(190, 116)

In [16]:
def extract_lists(s):
    try:
        two_strings = s.split(':')[1].split(' | ')
    except:
        print(s)
    res = []
    for s in two_strings:
        res.append([int(el.strip()) for el in s.split(' ') if el.strip() != ''])
    return res

extract_lists(lines[0])

[[2, 15, 17, 11, 64, 59, 45, 41, 61, 19],
 [4,
  36,
  62,
  43,
  94,
  41,
  24,
  25,
  13,
  83,
  97,
  86,
  61,
  90,
  67,
  7,
  15,
  58,
  18,
  19,
  38,
  17,
  49,
  52,
  37]]

In [19]:
total = 0
for l in lines:
    winning, given = extract_lists(l)
    points = 0
    # print(l)
    for el in given:
        if el in winning:
            if points == 0:
                points = 1
            else:
                points *= 2
            # print(el)
    total += points
total

15205

## Part 2
Just as you're about to report your findings to the Elf, one of you realizes that the rules have actually been printed on the back of every card this whole time.

There's no such thing as "points". Instead, scratchcards only cause you to win more scratchcards equal to the number of winning numbers you have.

Specifically, you win copies of the scratchcards below the winning card equal to the number of matches. So, if card 10 were to have 5 matching numbers, you would win one copy each of cards 11, 12, 13, 14, and 15.

Copies of scratchcards are scored like normal scratchcards and have the same card number as the card they copied. So, if you win a copy of card 10 and it has 5 matching numbers, it would then win a copy of the same cards that the original card 10 won: cards 11, 12, 13, 14, and 15. This process repeats until none of the copies cause you to win any more cards. (Cards will never make you copy a card past the end of the table.)

This time, the above example goes differently:
```
Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53
Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19
Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1
Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83
Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36
Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11
```

Card 1 has four matching numbers, so you win one copy each of the next four cards: cards 2, 3, 4, and 5.  
Your original card 2 has two matching numbers, so you win one copy each of cards 3 and 4.  
Your copy of card 2 also wins one copy each of cards 3 and 4.  
Your four instances of card 3 (one original and three copies) have two matching numbers, so you win four copies each of cards 4 and 5.   
Your eight instances of card 4 (one original and seven copies) have one matching number, so you win eight copies of card 5.  
Your fourteen instances of card 5 (one original and thirteen copies) have no matching numbers and win no more cards.  
Your one instance of card 6 (one original) has no matching numbers and wins no more cards.  
Once all of the originals and copies have been processed, you end up with 1 instance of card 1, 2 instances of card 2, 4 instances of card 3, 8 instances of card 4, 14 instances of card 5, and 1 instance of card 6. In total, this example pile of scratchcards causes you to ultimately have 30 scratchcards!  

Process all of the original and copied scratchcards until no more scratchcards are won. Including the original set of scratchcards, how many total scratchcards do you end up with?

In [32]:
def parse(s):
    try:
        card, values = s.split(':')
        two_strings = values.split(' | ')
    except:
        print(s)
    res = []
    for s in two_strings:
        res.append([int(el.strip()) for el in s.split(' ') if el.strip() != ''])

    card = int(card.split(' ')[-1].strip())
    return [card] + res

parse(lines[0])

[1,
 [2, 15, 17, 11, 64, 59, 45, 41, 61, 19],
 [4,
  36,
  62,
  43,
  94,
  41,
  24,
  25,
  13,
  83,
  97,
  86,
  61,
  90,
  67,
  7,
  15,
  58,
  18,
  19,
  38,
  17,
  49,
  52,
  37]]

In [54]:
from collections import deque

card_prize = {}
queue = deque(range(1, len(lines) + 1))
processed = 0

while queue:
    human_idx = queue.popleft()
    s = lines[human_idx-1]
    card_n, winning, given = parse(s)
    # print('processing card number', card_n)
    processed += 1
    
    if card_n in card_prize:
        matches = card_prize[card_n]
        # print('found card number', card_n, f"with {matches} matches") 
        for i in range(1, matches+1):
            queue.append(card_n+i)
            # print(f"Adding card {card_n+i} to queue") 
    else:
        matches = 0
        for el in given:
            if el in winning:
                matches += 1
                queue.append(card_n+matches)
                # print(f"Adding card {card_n+matches} to queue") 
        card_prize[card_n] = matches
    
processed

6189740

https://www.instagram.com/reel/CvDZ7nSrjCQ/?igshid=MzRlODBiNWFlZA%3D%3D